src/libpmemfile-posix/inode.c

```diff
@@ -0,0 +1,841 @@
+	vinode->inode = D_RW(inode);
+	if (inode_is_dir(vinode->inode) && parent) {
+		vinode->parent = vinode_ref(pfp, parent);
+
+		if (parent_refed)
+			*parent_refed = true;
+	}
+
+	if (parent && name && namelen)
+		vinode_set_debug_path_locked(pfp, parent, vinode, name,
+				namelen);
+
+	b->arr[empty_slot].pinode = inode;
+	b->arr[empty_slot].vinode = vinode;
+	map->inodes++;
+
+	if (is_new)
+		cb_push_front(TX_STAGE_ONABORT,
+			(cb_basic)vinode_unregister_locked,
+			vinode);
+
+end:
+	__sync_fetch_and_add(&vinode->ref, 1);
+	if (is_new && tx)
+		rwlock_tx_unlock_on_commit(&map->rwlock);
+	else
+		os_rwlock_unlock(&map->rwlock);
+
+	if (error)
+		errno = error;
+
+	return vinode;
+}
+
+/*
+ * inode_ref_new -- increases inode reference counter
+ *
+ * Assumes inode was allocated in the same transaction.
+ * Return volatile inode.
+ */
+struct pmemfile_vinode *
+inode_ref_new(PMEMfilepool *pfp,
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, true, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * inode_ref -- increases inode reference counter
+ *
+ * Assumes inode was not allocated in the same transaction.
+ * Return volatile inode.
+ */
+struct pmemfile_vinode *
+inode_ref(PMEMfilepool *pfp,
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, false, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * vinode_tx_unref -- decreases inode reference counter
+ *
+ * Must be called in a transaction.
+ */
+static bool
+vinode_tx_unref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_WORK);
+
+	if (__sync_sub_and_fetch(&vinode->ref, 1) > 0)
+		return false;
+
+	if (vinode->inode->nlink == 0) {
+		inode_array_unregister(pfp, vinode->orphaned.arr,
+				vinode->orphaned.idx);
+
+		inode_free(pfp, vinode->tinode);
+	}
+
+	return true;
+}
+
+/*
+ * vinode_unref -- decreases inode reference counter
+ *
+ * Can't be called in a transaction.
+ */
+void
+vinode_unref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_NONE);
+
+	struct pmemfile_inode_map *c = pfp->inode_map;
+
+	os_rwlock_wrlock(&c->rwlock);
+
+	while (vinode) {
+		struct pmemfile_vinode *to_unregister = NULL;
+		TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
+			struct pmemfile_vinode *parent = vinode->parent;
+
+			if (vinode_tx_unref(pfp, vinode))
+				to_unregister = vinode;
+
+			if (to_unregister && vinode != pfp->root)
+				vinode = parent;
+			else
+				vinode = NULL;
+		} TX_ONCOMMIT {
+			if (to_unregister)
+				vinode_unregister_locked(pfp, to_unregister);
+		} TX_ONABORT {
+			FATAL("!");
+		} TX_END
+	}
+
+	os_rwlock_unlock(&c->rwlock);
+}
+
+/*
+ * file_get_time -- sets *t to current time
+ */
+void
+file_get_time(struct pmemfile_time *t)
```

Comment:
how's this an inode related thing? It looks like a local helper function, but it's not.